sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -966,6 +952,22 @@ object ReplaceDistinctWithAggregate extends Rule[LogicalPlan] {
 }
 
 /**
+ * Replaces logical [[Intersect]] operator with a left-semi [[Join]] operator.
+ * {{{
+ *   SELECT a1, a2 FROM Tab1 INTERSECT SELECT b1, b2 FROM Tab2
+ *   ==>  SELECT a1, a2 FROM Tab1 LEFT SEMI JOIN Tab2 ON a1<=>b1 AND a2<=>b2
+ * }}}
+ */
+object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case Intersect(left, right) =>
+      assert(left.output.size == right.output.size)
+      val joinCond = left.output.zip(right.output).map { case (l, r) => EqualNullSafe(l, r) }
+      Join(left, right, LeftSemi, joinCond.reduceLeftOption(And))
```

Comment:
Does our current intersect do de-dup?






On Thu, Jan 7, 2016 at 8:50 AM -0800, "Xiao Li" <notifications@github.com> wrote:












In sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
> @@ -966,6 +952,22 @@ object ReplaceDistinctWithAggregate extends Rule[LogicalPlan] {
>  }
>  
>  /**
> + * Replaces logical [[Intersect]] operator with a left-semi [[Join]] operator.
> + * {{{
> + *   SELECT a1, a2 FROM Tab1 INTERSECT SELECT b1, b2 FROM Tab2
> + *   ==>  SELECT a1, a2 FROM Tab1 LEFT SEMI JOIN Tab2 ON a1<=>b1 AND a2<=>b2
> + * }}}
> + */
> +object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
> +  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
> +    case Intersect(left, right) =>
> +      assert(left.output.size == right.output.size)
> +      val joinCond = left.output.zip(right.output).map { case (l, r) => EqualNullSafe(l, r) }
> +      Join(left, right, LeftSemi, joinCond.reduceLeftOption(And))



You are right! It should remove duplicates. I just tried it in DB2. 



—
Reply to this email directly or view it on GitHub.